File: kalman/analyze_results.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import sys

def calculate_rmse(truth, est, columns):
    rmse = {}
    for col in columns:
        diff = truth[col] - est[col]
        rmse[col] = np.sqrt(np.mean(diff**2))
    return rmse

def analyze():
    # パス設定
    base_dir = os.path.dirname(os.path.abspath(__file__))
    truth_file = os.path.join(base_dir, 'GenerateData', 'truth_data.csv')
    est_file = os.path.join(base_dir, 'Results', 'estimation.csv')

    if not os.path.exists(truth_file):
        print(f"Error: {truth_file} not found.")
        return
    if not os.path.exists(est_file):
        print(f"Error: {est_file} not found.")
        return

    # データ読み込み
    truth_df = pd.read_csv(truth_file)
    est_df = pd.read_csv(est_file)

    # 時間合わせ（必要なら）
    # ここではサンプル数が同じで時間が同期していると仮定

    # カラムマッピング
    # truth: time, x, y, z, vx, vy, vz, roll, pitch, yaw
    # est: time, px, py, pz, vx, vy, vz, roll, pitch, yaw, ...
    
    # 位置
    pos_rmse = np.sqrt(np.mean((truth_df['x'] - est_df['px'])**2 + 
                               (truth_df['y'] - est_df['py'])**2 + 
                               (truth_df['z'] - est_df['pz'])**2))
    
    # 速度
    vel_rmse = np.sqrt(np.mean((truth_df['vx'] - est_df['vx'])**2 + 
                               (truth_df['vy'] - est_df['vy'])**2 + 
                               (truth_df['vz'] - est_df['vz'])**2))
    
    # 姿勢 (Yawの不連続性を考慮する必要があるが、ここでは単純差分)
    # wrapToPi 相当の処理が必要
    def angle_diff(a, b):
        d = a - b
        d = (d + 180) % 360 - 180
        return d

    roll_err = angle_diff(truth_df['roll'], est_df['roll'])
    pitch_err = angle_diff(truth_df['pitch'], est_df['pitch'])
    yaw_err = angle_diff(truth_df['yaw'], est_df['yaw'])

    att_rmse = np.sqrt(np.mean(roll_err**2 + pitch_err**2 + yaw_err**2))

    print(f"Position RMSE: {pos_rmse:.4f} m")
    print(f"Velocity RMSE: {vel_rmse:.4f} m/s")
    print(f"Attitude RMSE: {att_rmse:.4f} deg")

    # 個別のRMSE
    print(f"  Roll RMSE: {np.sqrt(np.mean(roll_err**2)):.4f} deg")
    print(f"  Pitch RMSE: {np.sqrt(np.mean(pitch_err**2)):.4f} deg")
    print(f"  Yaw RMSE: {np.sqrt(np.mean(yaw_err**2)):.4f} deg")

    # 発散チェック
    if pos_rmse > 1000 or np.isnan(pos_rmse):
        print("WARNING: Position estimation DIVERGED!")
    if att_rmse > 1000 or np.isnan(att_rmse):
        print("WARNING: Attitude estimation DIVERGED!")

    # グラフ作成
    plt.figure(figsize=(12, 10))
    
    plt.subplot(3, 1, 1)
    plt.plot(truth_df['time'], truth_df['x'], label='Truth X')
    plt.plot(est_df['time'], est_df['px'], label='Est X', linestyle='--')
    plt.title('Position X')
    plt.legend()
    plt.grid(True)

    plt.subplot(3, 1, 2)
    plt.plot(truth_df['time'], truth_df['roll'], label='Truth Roll')
    plt.plot(est_df['time'], est_df['roll'], label='Est Roll', linestyle='--')
    plt.title('Roll')
    plt.legend()
    plt.grid(True)

    plt.subplot(3, 1, 3)
    plt.plot(truth_df['time'], truth_df['yaw'], label='Truth Yaw')
    plt.plot(est_df['time'], est_df['yaw'], label='Est Yaw', linestyle='--')
    plt.title('Yaw')
    plt.legend()
    plt.grid(True)

    plt.tight_layout()
    plt.savefig(os.path.join(base_dir, 'Results', 'analysis_plot.png'))
    print("Plot saved to Results/analysis_plot.png")

if __name__ == "__main__":
    analyze()

File: kalman/analyze_results.m
% 結果解析スクリプト
fprintf('Loading estimation results...\n');
est = readtable('Results/estimation.csv');
truth = readtable('GenerateData/truth_data.csv');

% 基本統計
fprintf('\n=== Basic Statistics ===\n');
fprintf('Total samples: %d\n', height(est));
fprintf('Time range: %.2f - %.2f seconds\n', est.time(1), est.time(end));

% バイアス推定値の確認
fprintf('\n=== Bias Estimation ===\n');
fprintf('Accel bias (final): [%.6f, %.6f, %.6f] m/s^2\n', ...
    est.ba_x(end), est.ba_y(end), est.ba_z(end));
fprintf('Gyro bias (final): [%.6f, %.6f, %.6f] rad/s\n', ...
    est.bg_x(end), est.bg_y(end), est.bg_z(end));
fprintf('Gyro bias (final, deg/s): [%.6f, %.6f, %.6f]\n', ...
    rad2deg(est.bg_x(end)), rad2deg(est.bg_y(end)), rad2deg(est.bg_z(end)));

% バイアスが変化しているか確認
ba_max = max(abs([est.ba_x; est.ba_y; est.ba_z]));
bg_max = max(abs([est.bg_x; est.bg_y; est.bg_z]));
fprintf('Max accel bias magnitude: %.6f m/s^2\n', ba_max);
fprintf('Max gyro bias magnitude: %.6f rad/s (%.6f deg/s)\n', bg_max, rad2deg(bg_max));

% 誤差計算（初期化期間後）
init_samples = 2000;  % 5秒 @ 400Hz
idx = init_samples+1:height(est);

pos_err = sqrt((est.px(idx) - truth.x(idx)).^2 + ...
               (est.py(idx) - truth.y(idx)).^2 + ...
               (est.pz(idx) - truth.z(idx)).^2);
vel_err = sqrt((est.vx(idx) - truth.vx(idx)).^2 + ...
               (est.vy(idx) - truth.vy(idx)).^2 + ...
               (est.vz(idx) - truth.vz(idx)).^2);

% 姿勢誤差（unwrap処理）
roll_err = rad2deg(wrapToPi(deg2rad(est.roll(idx) - truth.roll(idx))));
pitch_err = rad2deg(wrapToPi(deg2rad(est.pitch(idx) - truth.pitch(idx))));
yaw_err = rad2deg(wrapToPi(deg2rad(est.yaw(idx) - truth.yaw(idx))));

fprintf('\n=== Estimation Errors (after initialization) ===\n');
fprintf('Position RMSE: %.4f m\n', rms(pos_err));
fprintf('Velocity RMSE: %.4f m/s\n', rms(vel_err));
fprintf('Roll RMSE: %.4f deg\n', rms(roll_err));
fprintf('Pitch RMSE: %.4f deg\n', rms(pitch_err));
fprintf('Yaw RMSE: %.4f deg\n', rms(yaw_err));

fprintf('\nPosition Max Error: %.4f m\n', max(pos_err));
fprintf('Velocity Max Error: %.4f m/s\n', max(vel_err));
fprintf('Roll Max Error: %.4f deg\n', max(abs(roll_err)));
fprintf('Pitch Max Error: %.4f deg\n', max(abs(pitch_err)));
fprintf('Yaw Max Error: %.4f deg\n', max(abs(yaw_err)));

% NaN/Infチェック
fprintf('\n=== Divergence Check ===\n');
has_nan = any(isnan(est.px) | isnan(est.py) | isnan(est.pz) | ...
              isnan(est.vx) | isnan(est.vy) | isnan(est.vz) | ...
              isnan(est.roll) | isnan(est.pitch) | isnan(est.yaw));
has_inf = any(isinf(est.px) | isinf(est.py) | isinf(est.pz) | ...
              isinf(est.vx) | isinf(est.vy) | isinf(est.vz));

if has_nan
    fprintf('WARNING: NaN detected in results!\n');
else
    fprintf('No NaN detected\n');
end

if has_inf
    fprintf('WARNING: Inf detected in results!\n');
else
    fprintf('No Inf detected\n');
end

% 判定
fprintf('\n=== Overall Assessment ===\n');
pass = true;
if rms(pos_err) > 5.0
    fprintf('FAIL: Position RMSE too high (%.4f > 5.0 m)\n', rms(pos_err));
    pass = false;
end
if rms(roll_err) > 5.0 || rms(pitch_err) > 5.0
    fprintf('FAIL: Attitude RMSE too high (Roll: %.4f, Pitch: %.4f)\n', ...
        rms(roll_err), rms(pitch_err));
    pass = false;
end
if has_nan || has_inf
    fprintf('FAIL: NaN or Inf detected\n');
    pass = false;
end

if pass
    fprintf('PASS: All checks passed!\n');
else
    fprintf('FAIL: Issues detected\n');
end
